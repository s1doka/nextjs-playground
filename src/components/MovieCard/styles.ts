import {makeStyles} from "@material-ui/core/styles";

export const useStyles = makeStyles(theme => ({
	root: {
		width: 150,
	},
	media: {
		width: '155px',
		height: '225px',
		borderRadius: 16
	},
	title: {
		fontWeight: 700,
		marginTop: 0,
		marginBottom: 0
	},
	releaseDate: {
		fontSize: 12
	}
}));
