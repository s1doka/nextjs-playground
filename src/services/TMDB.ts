import ky from "ky-universal";
import {
	APIQueryStrings,
	KeywordsResponse,
	MovieCreditsResponse,
	MovieDetailResponse,
	MovieImagesResponse,
	PopularMoviesResponse,
	PopularTVShowsResponse,
	TrendingMediaResponse,
	TrendingMediaType,
	TrendingTimeWindow,
} from "../types/TMDB";
import {
	MOVIE_CREDITS_ENDPOINT,
	MOVIE_DETAIL_ENDPOINT,
	MOVIE_IMAGES_ENDPOINT,
	MOVIE_KEYWORDS_ENDPOINT,
	POPULAR_MOVIES_ENDPOINT,
	POPULAR_TV_SHOWS_ENDPOINT,
	TMDB_HTTP_HEADERS,
	TRENDING_MEDIA_ENDPOINT,
} from "../configuration/settings";

async function getPopularMovies(
	searchParams?: APIQueryStrings,
): Promise<PopularMoviesResponse> {
	return await ky.get(
		POPULAR_MOVIES_ENDPOINT,
		{headers: TMDB_HTTP_HEADERS, searchParams},
	).json();
}

async function getPopularTVShows(
	searchParams?: Omit<APIQueryStrings, "region">,
): Promise<PopularTVShowsResponse> {
	return ky.get(
		POPULAR_TV_SHOWS_ENDPOINT,
		{headers: TMDB_HTTP_HEADERS, searchParams},
	).json();
}

async function getTrendingMedia(
	mediaType: TrendingMediaType,
	timeWindow: TrendingTimeWindow,
): Promise<TrendingMediaResponse> {
	return ky.get(
		`${TRENDING_MEDIA_ENDPOINT}/${mediaType}/${timeWindow}`,
		{headers: TMDB_HTTP_HEADERS},
	).json();
}

async function getMovieDetail(id: string): Promise<MovieDetailResponse> {
	return ky.get(`${MOVIE_DETAIL_ENDPOINT}/${id}`, {headers: TMDB_HTTP_HEADERS}).json();
}

async function getMovieKeywords(id: string): Promise<KeywordsResponse> {
	return ky.get(
		`${MOVIE_KEYWORDS_ENDPOINT.replace("$id", id)}`,
		{headers: TMDB_HTTP_HEADERS},
	).json();
}

async function getMovieCredits(id: string): Promise<MovieCreditsResponse> {
	return ky.get(
		`${MOVIE_CREDITS_ENDPOINT.replace("$id", id)}`,
		{headers: TMDB_HTTP_HEADERS},
	).json();
}

async function getMovieImages(id: string): Promise<MovieImagesResponse> {
	return ky.get(
		`${MOVIE_IMAGES_ENDPOINT.replace("$id", id)}`,
		{headers: TMDB_HTTP_HEADERS},
	).json();
}

function formatReleaseDate(releaseDate: string): string {
	return new Intl.DateTimeFormat(
		"en",
		{timeZone: "UTC", day: "2-digit", month: "long", year: "numeric"},
	).format(new Date(releaseDate));
}

function formatRating(voteAverage: number): string {
	return `${voteAverage * 10}`;
}

function buildReleaseImagePath(releasePosterPath: string) {
	return `https://image.tmdb.org/t/p/w500${releasePosterPath}`;
}

export {
	buildReleaseImagePath,
	formatRating,
	formatReleaseDate,
	getMovieCredits,
	getMovieDetail,
	getMovieImages,
	getMovieKeywords,
	getPopularMovies,
	getPopularTVShows,
	getTrendingMedia,
};
