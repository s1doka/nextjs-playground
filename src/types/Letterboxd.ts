// A cursor is a string value provided by the API. It should be treated as an opaque value — don’t change it.
export type Cursor = string;

export type AuthenticationTokenGrantType = "password";

export type AuthenticationTokenRequest = {
	grantType: AuthenticationTokenGrantType;
	username: string;
	password: string;
};

export type Image = {
	sizes: Array<ImageSize>;
};

export type ImageSize = {
	width: number;
	height: number;
	url: string;
};

export type NewsItem = {
	title: string;
	image: Image;
	url: string;
	shortDescription: string;
	longDescription: string;
};

export type NewsRequest = {
	cursor: Cursor;
	perPage: number;
};
