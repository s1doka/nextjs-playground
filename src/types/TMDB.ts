/**
 * Paginated response from the API.
 *
 * @typeParam T - Specify the type for the results list.
 */
export type PaginatedResponse<T> = {
	/*
   * The actual page.
   * */
	page: number;
	/*
   * Number of total results.
   * */
	total_results: number;
	/*
   * Number of total pages.
   * */
	total_pages: number;
	/*
   * List of results.
   * */
	results: Array<T>;
};

/**
 * Available query strings to filter results.
 */
export type APIQueryStrings = {
	/*
   * Pass a ISO 639-1 value to display translated data for the fields that support it.
   * */
	language: string;
	/*
   * Specify which page to query.
   * */
	page: number;
	/*
   * Specify a ISO 3166-1 code to filter release dates. Must be uppercase.
   * */
	region: string;
};

export type Movie = {
	poster_path: string | null;
	adult: boolean;
	overview: string;
	release_date: string;
	genre_ids: Array<number>;
	id: number;
	original_title: string;
	original_language: string;
	title: string;
	backdrop_path: string | null;
	popularity: number;
	vote_count: number;
	video: boolean;
	vote_average: number;
};

export type MovieDetail = Movie & {
	belongs_to_collection: object | null;
	budget: number;
	genres: Array<Genre>;
	homepage: string | null;
	imdb_id: string | null;
	production_companies: Array<ProductionCompany>;
	production_countries: Array<ProductionCountry>;
	revenue: number;
	runtime: number | null;
	spoken_languages: Array<SpokenLanguage>;
	status: MovieStatus;
	tagline: string | null;
};

export type MovieStatus =
	| "Rumored"
	| "Planned"
	| "In Production"
	| "Post Production"
	| "Released"
	| "Canceled";

export type Genre = {
	id: number;
	name: string;
};

export type ProductionCompany = {
	id: number;
	name: string;
	logo_path: string | null;
	origin_country: string;
};

export type SpokenLanguage = {
	iso_639_1: string;
	name: string;
};

export type ProductionCountry = {
	iso_639_1: string;
	name: string;
};

export type TVShow = {
	poster_path: string | null;
	popularity: number;
	id: number;
	backdrop_path: string | null;
	vote_average: number;
	overview: string;
	first_air_date: string;
	origin_country: Array<string>;
	genre_ids: Array<number>;
	original_language: string;
	vote_count: number;
	name: string;
	original_name: string;
};

export type TrendingMediaType = "all" | "movie" | "person" | "tv";

export type TrendingTimeWindow = "day" | "week";

export type PopularMoviesResponse = PaginatedResponse<Movie>;

export type PopularTVShowsResponse = PaginatedResponse<TVShow>;

export type TrendingMediaResponse = PaginatedResponse<Movie & TVShow>;

export type MovieDetailResponse = MovieDetail;
