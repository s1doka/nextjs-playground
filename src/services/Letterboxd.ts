import ky from "ky-universal";
import {AuthenticationTokenRequest, NewsRequest} from "../types/Letterboxd";

const LETTERBOXD_HOST = process.env.NEXT_PUBLIC_LETTERBOXD_HOST;

function createAuthenticationTokenFormData(formValues: object): FormData {
	const formData = new FormData();

	Object.keys(formValues).forEach((key) => formData.append(key, formValues[key]));

	return formData;
}

async function generateAuthenticationToken(
	{grantType, password, username}: AuthenticationTokenRequest,
) {
	const body = createAuthenticationTokenFormData({grantType, password, username});

	try {
		const response = ky.post(`${LETTERBOXD_HOST}/auth/token`, {body});
		return response;
	} catch (error) {
		console.log({error});
	}
}

async function getNews(params?: NewsRequest) {
	try {
		return ky.get(`${LETTERBOXD_HOST}/news`).json();
	} catch (error) {
		console.log({error});
	}
}

export {generateAuthenticationToken, getNews};
