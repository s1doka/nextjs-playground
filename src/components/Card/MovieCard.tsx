import {useStyles} from "./styles";
import {formatReleaseDate} from "../../services/TMDB";
import {Movie} from "../../types/TMDB";

type Props = {
	movie: Movie;
};

function MovieCard({movie}: Props) {
	const classes = useStyles();

	return <div className={classes.root}>
		<img className={classes.media}
		src={`https://image.tmdb.org/t/p/w500${movie.poster_path}`}
		alt={movie.original_title} />
		<p className={classes.title} data-testid={`movie-title-${movie.id}`}>
			{movie.original_title}
		</p>
		<span className={classes.releaseDate} data-testid={`movie-release-date-${movie.id}`}>
			{formatReleaseDate(movie.release_date)}
		</span>
	</div>;
}

export default MovieCard;
