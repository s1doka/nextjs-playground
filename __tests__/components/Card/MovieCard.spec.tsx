import {render} from "../../test-utils";

import MovieCard from "../../../src/components/Card/MovieCard";
import {Movie} from "../../../src/types/TMDB";
import {formatReleaseDate} from "../../../src/services/TMDB";

describe(
	"<MovieCard/>",
	() => {
		const movie: Movie = {
			"poster_path": "/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg",
			"adult": false,
			"overview": "From DC Comics comes the Suicide Squad, an antihero team of incarcerated supervillains who act as deniable assets for the United States government, undertaking high-risk black ops missions in exchange for commuted prison sentences.",
			"release_date": "2016-08-03",
			"genre_ids": [14, 28, 80],
			"id": 297_761,
			"original_title": "Suicide Squad",
			"original_language": "en",
			"title": "Suicide Squad",
			"backdrop_path": "/ndlQ2Cuc3cjTL7lTynw6I4boP4S.jpg",
			"popularity": 48.261451,
			"vote_count": 1_466,
			"video": false,
			"vote_average": 5.91,
		};

		test(
			"snapshot testing",
			() => {
				const {asFragment} = render(<MovieCard movie={movie} />);
				expect(asFragment()).toMatchSnapshot();
			},
		);

		test("should render the movie title and release date", () => {
			const { getByTestId}= render(<MovieCard movie={movie} />)

			expect(getByTestId(`movie-title-${movie.id}`)).toBeVisible()
			expect(getByTestId(`movie-title-${movie.id}`)).toHaveTextContent(movie.original_title)

			expect(getByTestId(`movie-release-date-${movie.id}`)).toBeVisible()
			expect(getByTestId(`movie-release-date-${movie.id}`)).toHaveTextContent(formatReleaseDate(movie.release_date))
		})
	},
);
