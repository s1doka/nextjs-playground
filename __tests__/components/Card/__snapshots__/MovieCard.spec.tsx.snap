// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard/> snapshot testing 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-largeCard-1"
  >
    <div
      class="makeStyles-mediaWrapper-5"
    >
      <img
        alt="Suicide Squad"
        class="makeStyles-largeImage-3"
        src="https://image.tmdb.org/t/p/w500/e1mjopzAS2KNsvpbpahQ1a6SkSn.jpg"
      />
      <div
        class="makeStyles-rating-10"
      >
        59.1
        <span
          class="makeStyles-ratingPercentage-11"
        >
          %
        </span>
      </div>
    </div>
    <div
      class="makeStyles-info-9"
    >
      <p
        class="makeStyles-title-7"
        data-testid="movie-title-297761"
      >
        Suicide Squad
      </p>
      <span
        class="makeStyles-releaseDate-8"
        data-testid="movie-release-date-297761"
      >
        August 03, 2016
      </span>
    </div>
  </div>
</DocumentFragment>
`;
